Find compatible GradleJavaHome and notify client in case of incompatibility

Tanish-Ranjan opened the pull request:
Addressing [issue 75](https://github.com/microsoft/build-server-for-gradle/issues/75) and [issue 76](https://github.com/microsoft/build-server-for-gradle/issues/76)

Tanish-Ranjan committed:
Notify Client on Java Home incompatibility
Changes:
- Added a ClientNotifier.java, a utility class for sending notifications to the client.
- LifecycleService now keeps a reference to the BuildClient for sending notifications.
- Updated Launcher.java to set BuildClient on LifecycleService.java
- Added method in utils.java to return the lease compatible Java version.
- Added unit tests in BuildTargetServerIntegrationTest.java to check if the compatibility notifications are returned properly.

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -155,6 +164,13 @@ private void updateGradleJavaHomeIfNecessary(URI rootUri) {
         if (jdkInstallation != null) {
           preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+        } else {
+          new ClientNotifier(client).sendNotification(
```
Is it possible to expand the compatibility check scope?

Some cases for consideration:

1. If user specifies Gradle Java Home in preference, is it a valid Java home? Is it compatible with the gradle version being used?
2. If user does not specify the gradle version, we still can infer the gradle version used for the project, can we do the compatibility check?

Tanish-Ranjan replied:
Commit [8731678](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/87316784aaf454d061a832bec0a38694301f8149) address this. We are now checking for `GradleJavaHome`, if it is available and compatible. If not we move to find a compatible JDK from the list of JDKs we get from preferences.

Tanish-Ranjan committed:
GradleJavaHome detection for compilation of Java programs.

Changes:
- Added method to retrieve the GradleJavaHome for the given project.
- Added a new utility class JavaUtils. It contains methods to extract the java version from the JDK file and check java version compatibility.
- Added new test class JavaUtilsTest containing unit tests for methods in JavaUtils.
- Updated LifecycleService to check for GradleJavaHome compatibility before checking JDKs in the preferences.

Tanish-Ranjan committed:
Patch

Changes:
- Updated LifecycleService#getJdkToLaunchDaemon to use feature release for check with selected version.
- Refactored JavaVersion variables for consistency in naming.
- Added unit test for retrieving the oldest compatible java version.
- Updated tests for UserJavaHome to prevent the use of GradleJavaHome for receiving the correct error message on client.

Tanish-Ranjan commented:
To check if the selected JDK is compatible or not we are now comparing the feature release of the Java Version. Previously, if we retrieved the latest compatible JDK version of say `21` and the client has JDK version `21.0.1`. Then `Runtime.Version#compareTo` it would say the client JDK version is greater than the highest compatible Java version. So I have resorted to using the feature release for compatibility check.

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java:
```diff
@@ -51,23 +51,40 @@ public GradleApiConnector(PreferenceManager preferenceManager) {
+   * @param projectUri URI of the project to get the gradle java home for.
+   * @return GradleJavaHome of the project or null upon failure.
+   */
+  public File getGradleJavaHome(URI projectUri) {
```
This method will return the Java home which used to launch build server itself, do I understand correctly?

Tanish-Ranjan replied:
No, I don't think so. This method gets `GradleJavaHome` for the given gradle project from the [BuildEnvironment](https://docs.gradle.org/current/javadoc/org/gradle/tooling/model/build/BuildEnvironment.html) > [JavaEnvironment](https://docs.gradle.org/current/javadoc/org/gradle/tooling/model/build/JavaEnvironment.html). 

Gradle gets the `GradleJavaHome` property in the following [priority order](https://docs.gradle.org/current/userguide/build_environment.html#priority_for_configurations):
1. Command Line arg
2. Gradle Properties File
3. Environment Variable

In our case only the last two will be valid so, this method will return the `GradleJavaHome` defined in the `gradle.properties` file, otherwise it will fallback to using the `JAVA_HOME` environment variable for the JDK location.

jdneo replied:
Thank you, I learnt from it!

Just a small suggestion: would you like to do some verification about the behavior? Because I found that some time, the behavior from command line is different from calling the TAPI. Just in case...

Tanish-Ranjan replied:
Okay. I will verify it.

Tanish-Ranjan replied:
Continued [here](https://github.com/microsoft/build-server-for-gradle/pull/165#issuecomment-2232644129).

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,114 @@ public void exit() {
+    }
+
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
```
Do we know when `gradleVersion` will be null?

Tanish-Ranjan replied:
I think you meant empty? The answer is yes, when retrieving the Gradle version using `GradleApiConnector#getGradleVersion` from the `BuildEnvironment` and a `BuildException` gets thrown. In this case we will get an empty string.

jdneo replied:
Do you know when will `BuildEnvironment` get thrown? Just asking for curiosity.

Tanish-Ranjan replied:
The check already existed in the codebase. Although [BuildExceptions](https://docs.gradle.org/current/javadoc/org/gradle/tooling/BuildException.html) are thrown when a model cannot be built. Based on [this](https://stackoverflow.com/questions/53991804/could-not-fetch-model-of-type-buildenvironment-using-gradle-distribution) I would say building the `BuildEnvironment` model would throw an exception due to incorrectly defined Gradle properties.

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/utils/JavaUtils.java:
```diff
@@ -0,0 +1,79 @@
+  public static String getJavaVersionFromFile(File jdkFile) throws IOException {
+
+    ProcessBuilder processBuilder =
+        new ProcessBuilder(jdkFile.getAbsolutePath() + "/bin/java", "-version");
```
what will happen if the executable file does not exist?

Tanish-Ranjan replied:
That should ideally throw an `IOException` when we try to start the process using `processBuilder.start()`. If the exception occurs it will be up to the class calling this utility function to handle it.

I think we can have more unit tests for this method, right now we are only simulating the error in `testGetJavaVersionFromFileForException`.

jdneo replied:
Ok, let's add one for file not exist case

Tanish-Ranjan replied:
Added in commit [1e1ad04](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/1e1ad040873483729109c47bc7f489b24f0afa67)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,114 @@ public void exit() {
     }
+
+    // Notify client, for no compatible JDK can be found
+    new ClientNotifier(client).sendNotification(
```
I'm thinking that if this is over design at this stage for the `ClientNotifier`? Does that mean we need to create a new notifier instance every time we want to send some msg to the client?

I like the idea of wrap things like taskid & originId into it. Maybe at current stage, just call the method from `client` directly? In the future, if we found that this is not enough, we can refactor to introduce sth like the notifier. WDYT?

Tanish-Ranjan replied:
If there is no requirement for the `taskId` and `originId` at this stage, I could make the `sendNotification` method static and there won't be the need to create new instances of ClientNotifier.

jdneo replied:
I would prefer directly using `client`.

Tanish-Ranjan replied:
Lifecycle service directly uses the client to send notifications with [847c8a2](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/847c8a25e0d409e03f2628beea5aa87f27c8dbd2)

Tanish-Ranjan committed:
Refactored ClientNotifier to flatten sendNotification calls.

Changes:
- Removed methods for adding originId and taskId to the notification in ClientNotifier.
- Refactored sendNotification to be static and accept the BuildClient directly in the parameter.
- Added a new test case to JavaUtilsTest verifying if IOException is properly thrown in case of non-existent executable.

Tanish-Ranjan committed:
Updated LifecycleService to directly send notification instead of delegating to ClientNotifier.

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java:
```diff
@@ -51,23 +51,40 @@ public GradleApiConnector(PreferenceManager preferenceManager) {
+  private BuildEnvironment getBuildEnvironment(ProjectConnection connection) {
+    return connection
         .model(BuildEnvironment.class)
         .withArguments("--no-daemon")
```
`--no-daemon` does not have effect when using tooling api

Tanish-Ranjan replied:
Yes, I saw it get merged in #172. When this PR is ready for merge I will merge the last commit from the develop branch into this, resolving any merge conflicts. This should get fixed there.

jdneo replied:
Or you can just rebase the PR based on the develop branch

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java:
```diff
@@ -153,18 +170,20 @@ public StatusCode runTasks(URI projectUri, ProgressReporter reporter, String...
     ProgressReporter reporter = new DefaultProgressReporter(client);
     try (ProjectConnection connection = getGradleConnector(projectUri).connect()) {
-      String gradleVersion = getGradleVersion(connection);
+      String gradleVersion = getGradleVersion(projectUri);
```
Could you reuse the connection here? For example, overloading the `getGradleVersion()`?

Tanish-Ranjan replied:
Addressed in commit [fbf550b](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/fbf550b288559f5f92a474307885ce56e417d962)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,117 @@ public void exit() {
-   * </ul>
-   *
-   * <p>The JDK installation path string will be set to {@link Preferences#gradleJavaHome}.
+   * Finds a suitable version of JDK to use for compilation of Java programs.
```
Let's say a JDK compatible with the target Gradle version. Because it's the jdk which is used for Gradle operations (for example running tasks or acquiring model information). It's not only for compilation.

Tanish-Ranjan replied:
Addressed in commit [fbf550b](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/fbf550b288559f5f92a474307885ce56e417d962)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,117 @@ public void exit() {
-        if (jdkInstallation != null) {
-          preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+      // Use GradleJavaHome if compatible
+      if (gradleJavaHome != null) {
```
Will `gradleJavaHome` null here? Given we have logic at line 152

Tanish-Ranjan replied:
Yes, it will be null if the `GradleApiConnector#getGradleJavaHome` gets a `BuildException`. In this case we will get a null value.

jdneo replied:
Ok, my next question is that will `GradleApiConnector#getGradleJavaHome` gurantee that the found java home is compatible?

Tanish-Ranjan replied:
No. It is only going to return the `GradleJavaHome`, looking for it in the priority order listed in [gradle docs](https://docs.gradle.org/current/userguide/build_environment.html#priority_for_configurations). It only queries for a `GradleJavaHome` and does not check for compatibility. It can even return non existent filepath if the user has wrong configurations for the project (eg. non existent jdk path in gradle.properties or JAVA_HOME).

Tanish-Ranjan committed:
Addressing comments

Tanish-Ranjan commented:
The JDK lookup for gradle seems to be different from the way it is listed in the [docs](https://docs.gradle.org/current/userguide/build_environment.html#priority_for_configurations).

Priority order I found out:
1. Command Line args
2. Gradle Properties
3. Installed JDK (gets from the registry I think)
4. JAVA_HOME (Env variable)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,117 @@ public void exit() {
+
+      // Pick a compatible JDK from the JDKs available in Preferences
+      if (preferences.getJdks() != null && preferences.getJdks().isEmpty()) {
+        return getJdkToLaunchDaemon(
```
If the preference has set some jdks but no one is compatible with gradle, since we return here, the client will not be notified.

Tanish-Ranjan replied:
Addressed in commit [c7f5d83](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/c7f5d838683fa585efa8d5ffb25ebe9b87d112ea)

Tanish-Ranjan committed:
Addressing comments

Tanish-Ranjan committed:
Addressing checkstyle warnings

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,120 @@ public void exit() {
-      if (Version.parse(javaVersion).compareTo(Version.parse(highestJavaVersion)) <= 0
-          && (selected == null || Version.parse(selected.getKey())
-              .compareTo(Version.parse(javaVersion)) < 0)) {
+      Boolean isSelectedVersionHigher =
```
nit:

1. There is no need to use object Boolean here, just primitive `boolean` should be fine.
2. Seems that name the variable to `isHigherThanSelected` makes more sense.

Tanish-Ranjan replied:
Addressed in commit [e7de6d8](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/e7de6d870c1893f46e4d25f1bfb2b59ffed35218)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,120 @@ public void exit() {
+      if (gradleJavaHome != null) {
+
+        try {
+          String gradleJavaHomeVersion = JavaUtils.getJavaVersionFromFile(gradleJavaHome);
```
What will happen if, `JavaUtils.getJavaVersionFromFile` returns a string which is not a purely string? Will it throw exception in the `JavaUtils.isCompatible()`? Will the exception be caught?

Tanish-Ranjan replied:
What do you mean by purely string? `JavaUtils#getJavaVersionFromFile` returns a string containing the version of the jdk or `null` upon failure. I will add a null check to prevent `NullPointerException` in `JavaUtils#isCompatible`.

jdneo replied:
Sorry, my typo. 

I mean, for example, assume `JavaUtils.getJavaVersionFromFile` returns "foo17.0.1"

Tanish-Ranjan replied:
Is that a special case in which the version string contains additional characters?
For example if we run the command `java -version`, the supposed output would be in the given format correct?
```
java -version
java version "21.0.3" 2024-04-16 LTS
```
And now we are extracting the version printed between the double quotes.

jdneo replied:
I'm not sure. Since you know there are so many different java distribution, I'm not sure if they will all print the version like that.

I mean. no need to cover all those cases. If we found that the version string cannot be correctly parsed, just makes sure the server will not fail down due to the unhandled exceptions.

Tanish-Ranjan replied:
Addressed in commit [e7de6d8](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/e7de6d870c1893f46e4d25f1bfb2b59ffed35218)

Tanish-Ranjan committed:
Addressing comments

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java:
```diff
@@ -119,61 +134,121 @@ public void exit() {
+    if (StringUtils.isNotEmpty(filePath)) {
+      gradleJavaHome = new File(filePath);
+    } else {
+      gradleJavaHome = connector.getGradleJavaHome(rootUri);
```
Some new thoughts here:

One assumption: According to my past experience, most of the users will use default value, and the default value for the gradle java home and gradle version are empty.

That means, the code here will call tooling api to get the `BuildEnvironment` twice. This can be improved by just get the `BuildEnvironment` once, and later, get the gradle and java version on demand.

Tanish-Ranjan replied:
Okay. We can save the last queried `BuildEnvironment` and use that if needed again. Even though the `BuildEnvironment` is project dependent, we are only using it for the root project. I will mention the use case in the javaDoc for future references.

jdneo replied:
Do you mean cache the `BuildEnvironment` in the connector?

Tanish-Ranjan replied:
Yes in the `GradleApiConnector#getBuildEnvironment` we can cache the `BuildEnvironment` obtained and then return it instead for the next calls. We may add a parameter to this method as well to force it to fetch a new `BuildEnvironment` if needed.

jdneo replied:
Might be a little bit over-design. Just get it once here is good enough

Tanish-Ranjan replied:
Fixed in commit [04ed182](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/04ed182c48bf04e766b4189900cdfbb470550b50)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/utils/JavaUtils.java:
```diff
@@ -0,0 +1,79 @@
+   * @return true if given javaVersion is less than or equal to the latestJavaVersion
+   *        and greater than or equal to the oldestJavaVersion.
+   */
+  public static Boolean isCompatible(
```
nit: Use `boolean`

Tanish-Ranjan replied:
Fixed in commit [04ed182](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/04ed182c48bf04e766b4189900cdfbb470550b50)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/utils/JavaUtils.java:
```diff
@@ -0,0 +1,79 @@
+      String latestJavaVersion
+  ) {
+
+    Version versionToCheck = Version.parse(javaVersion);
```
catch the `IllegalArgumentException` and `NumberFormatException`

consider to add null check as well

Tanish-Ranjan replied:
I didn't add those checks there because it is inside a utility class, and there are often instances where the caller would like to handle the exception on its own. So we may just throw the error there and handle the exception at the call site.

jdneo replied:
Ok, just make sure there is no uncaught exception.

Tanish-Ranjan replied:
Fixed in commit [04ed182](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/04ed182c48bf04e766b4189900cdfbb470550b50)

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/utils/JavaUtils.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class JavaUtils {
+
+  public JavaUtils() {
```
should be private?

Tanish-Ranjan replied:
Fixed in commit [04ed182](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/04ed182c48bf04e766b4189900cdfbb470550b50)

jdneo commented on server/src/test/java/com/microsoft/java/bs/core/internal/server/BuildTargetServerIntegrationTest.java:
```diff
@@ -477,6 +535,87 @@ void testCompilingSingleProjectServer() {
   }
 
+  @Test
+  void testIncompatibleUserJavaHomeProjectServer() {
```
Makes more sense to put it into `LifecycleServiceIntegrationTests`?

Tanish-Ranjan replied:
Are you proposing to create a new test file? I didn't find any existing test file called `LifecycleServiceIntegrationTests`.

jdneo replied:
Yes it is.

Tanish-Ranjan replied:
Fixed in commit https://github.com/microsoft/build-server-for-gradle/pull/165/commits/8ed01255f35de1f7ab772357cfc30e44a3f26174

Tanish-Ranjan committed:
Broke down the logic for selection of java home into smaller functions and flattened the logic.

Changes:
- Updated JavaDocs in GradleApiConnector.
- Removed getGradleJavaHome from GradleApiConnector.
- Added getBuildEnvironment method to extract the BuildEnvironment variable directly from the GradleApiConnector.
- Made JavaUtils constructor private and added throws and @Nonnull annotations to prevent unwanted exceptions.
- Broke down and flattened the logic for getSuitableJdk from LifecycleService into two separate methods.
- Added getGradleVersion method in LifecycleService to extract the gradle version for the given project.
- Updated getSuitableJdk to getGradleCompatibleJdk to contain only the java home selection logic.
- Added a new method setGradleJavaHome which calls the previous methods to select and set the java home and send notification to the client upon failure.

Tanish-Ranjan committed:
Separated BuildTargetServerIntegrationTest into two separate tests.

Changes:
- Removed BuildTargetServerIntegrationTest
- Added IntegrationTest. Base class for integration tests containing the setup.
- Added BuildTargetServiceIntegrationTest. Subclass of IntegrationTest responsible for handling build target integration tests.
- Added LifecycleServiceIntegrationTest. Subclass of IntegrationTest responsible for handling lifecycle integration tests.

jdneo commented:
Hi @Tanish-Ranjan, The PR has some conflicts with the develop branch now. Could you please resolve them?

jdneo commented on server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java:
```diff
@@ -53,22 +53,63 @@ public GradleApiConnector(PreferenceManager preferenceManager) {
+   */
+  public BuildEnvironment getBuildEnvironment(URI projectUri) {
+    try (ProjectConnection connection = getGradleConnector(projectUri).connect()) {
+      return connection
```
return getBuildEnvironment(connection);

Tanish-Ranjan replied:
Fixed in https://github.com/microsoft/build-server-for-gradle/pull/165/commits/7e4749a704b01421c5b6282ea0204d8856b32636

jdneo commented on server/src/test/java/com/microsoft/java/bs/core/internal/server/IntegrationTest.java:
```diff
@@ -0,0 +1,431 @@
+        capabilities);
+  }
+
+  protected static InitializeBuildParams getInitializedBuildParamsWithJdks(
```
Put this as a member of LifecycleServiceIntegrationTest?

Tanish-Ranjan replied:
Fixed in https://github.com/microsoft/build-server-for-gradle/pull/165/commits/7e4749a704b01421c5b6282ea0204d8856b32636

jdneo commented on server/src/test/java/com/microsoft/java/bs/core/internal/server/IntegrationTest.java:
```diff
@@ -0,0 +1,431 @@
+          + testName.getMethodName() + "," + getTestNameHierarchy(testName);
+    }
+
+    TestStartEx getTestStart(String suiteName, String className, String methodName,
```
Since this is the base class, I think it should only provide the minimum required members. These test related method better move to the `BuildTargetServiceIntegrationTest`?

Tanish-Ranjan replied:
Fixed in https://github.com/microsoft/build-server-for-gradle/pull/165/commits/7e4749a704b01421c5b6282ea0204d8856b32636

Tanish-Ranjan replied:
Fixed in https://github.com/microsoft/build-server-for-gradle/pull/165/commits/7e4749a704b01421c5b6282ea0204d8856b32636

Tanish-Ranjan committed:
Refactor IntegrationTests

Changes:
- Moved methods from IntegrationTest to LifecycleService and BuildTargetService IntegrationTest to keep only the common methods needed by both integration tests.

jdneo commented on server/src/test/java/com/microsoft/java/bs/core/internal/server/LifecycleServiceIntegrationTest.java:
```diff
@@ -0,0 +1,137 @@
+        "0.1.0",
+        root.toURI().toString(),
+        capabilities
+    );
```
Looks like we can directly get the initParams by `IntegrationTest.getInitializeBuildParams(projectDir)`, which can remove duplicated code

Tanish-Ranjan replied:
Fixed in https://github.com/microsoft/build-server-for-gradle/pull/165/commits/24fdf827bf77e943ea93778a3b3fdd7b78a9b94c

Tanish-Ranjan committed:
Remove duplicate code.

Tanish-Ranjan committed:
Checkstyle fix

Tanish-Ranjan committed:
Refactored Java Home detection to prioritize gradle default behavior compatibility.

Changes:
- Added new method to check compatibility of Java Home with a probe build in GradleApiConnector
- Refactored LifecycleService#setGradleJavaHome to prioritize the default gradle configuration and send a notification to the client if the default behavior is changed.
- Added new test for java home detection with gradle default behavior.

Tanish-Ranjan commented:
Prioritizing Gradle default behavior for java home detection with https://github.com/microsoft/build-server-for-gradle/pull/165/commits/3b6888733339c8f42229b18fa9cecfa09f84d500

reinsch82 reviewed (approved):
Looks good to me.

thx

jdneo reviewed (approved):
LGTM
